Chromosome view yonah

y330 opened the pull request:
For yukthi to look at:
Eplant/views/ChromosomeViewer/Viewer/GeneList.Tsx Line 111

y330 committed:
first commit > not sure how to setup getInitialData to fetch the chromosomes, using the species state, and provide the chromosomes as a prop to the component

y330 committed:
- started the get initial data function, but
the fetch is not working to get the chromosomes
- hardcoded the chromosome data to the component so i could at least see
the view working, and the getGenes fetch function is working fine
- so there must be something wrong with my getintitialdata function

y330 committed:
=====changes=====
=================
-[X] styled: genelist popup box, start and end of the range based on click, padding of items on the genelist
-[X] added: gene info popup(visible on clicking gene in genelist popup box), chromosome name and size labels
TODO
-[] add lines starting at the top and bottom popup box that converge onto the specific location of the chromsome when clicked
-[] fix outside click of gene info popup closes the all popups, and when you click inside the geneinfo popup it closes TODO
-[] fix getInitialData so that it fetches the chromosomes, currently just using hardcoded chromosome object
-[] tweak more styling

y330 committed:
fixed the gene info popups somewhat, they still
need some work

y330 committed:
- Moved GeneInfoDialog to GeneList component, and now functionality is much better, made info popup draggable, possible to interact with without closing and more

y330 committed:
- Styled the GeneInfoPopup to include icons, and action button
- fixed GeneInfoPopup so thats its draggable and can be opened and closed properly
- TODO
 + implement fetchGeneSummary function in Chromosome.jsx to get active gene location from geneticElement prop
 + implemnet setActiveGene in GeneList and GeneInfoPopup to allow the user to load genes into the collection on the sidebar

y330 committed:
- implemented eplant panzoom component in base chromosome viewer/index.tsx instead of using a third party library
- added view actions
- touched up the view functions and made it more functional.
- added functionality for changing the gap between chromosomes depending on how many there are. i.e. if its Populus_trichocarpa

y330 committed:
- changed geneList popup to have ability to be clicked outside it
- added gene aliases to genelist popup
- switched back to MapInteractionCSS component for zooming because panzoom was causing bugs with the gene info popup while dragging
- ran into the same fatal errors as last time while trying to implement showing the active gene location on the chromosome, so had to hardcode the chromosomes in the component function again

y330 committed:
- Added snackbar-like popup that shows up when clicking genes to prompt user to open gene info popup

y330 committed:
Added snackbarlike popup that shows up when clicking gene to prompt user to open gene info popup, this fixed the double clicking glitch on the genelist

y330 committed:
removed snackbar, fixed geneinfo popup state
implmented load gene function
tried implementing showing the location of activegene on the chromosome,but ran into promise errors for some reason

y330 committed:
Added activeGeneArrow to show where the active gene is on the chromosome ISSUES: it mvoes wwith the screen as tyou pan through the chromosomes
PREVIOUS COMMIT: forgot to mention that I Added load selected gene into collections feature, but it has an issue where it only updates the geneticElements state when searching from the searchbar, but the feature still works
changed the zoom in/ out buttons on the map interaction css component(zoom/pan)

y330 committed:
Implmented active gene arrow to show where selected gene is on the Chromosome
TODO:
- make it so it shows all the genes in the collection
- style gene info popup
- add logic to determine if gene arrow will be on left or right of chromosome based on strand being + or -
- fix gene list popup placement

y330 committed:
deletedsome imports

y330 committed:
made seperate GeneInfoPopup component

Yukthiw commented on Eplant/views/ChromosomeViewer/Viewer/GeneList.tsx:
```diff
@@ -0,0 +1,247 @@
+
+	// Other/Global State
+	const [activeGeneId, setActiveGeneId] = useActiveGeneId()
+	const geneticElements = useGeneticElements()
```
```suggestion
	const [geneticElements, setGeneticElements] = useGeneticElements()
```
All Jotai hooks are used similar to useState, in this case you don't need `useSetGeneticElements()` and should just use the setter provided by `useGeneticElements`

Yukthiw commented on Eplant/views/ChromosomeViewer/Viewer/GeneList.tsx:
```diff
@@ -0,0 +1,247 @@
+	// Other/Global State
+	const [activeGeneId, setActiveGeneId] = useActiveGeneId()
+	const geneticElements = useGeneticElements()
+	const setGeneticElements = useSetGeneticElements()
```
```suggestion
```

Yukthiw commented on Eplant/views/ChromosomeViewer/Viewer/GeneList.tsx:
```diff
@@ -0,0 +1,247 @@
+				selectedGene.aliases
+			)
+			geneticElements[0].push(gene)
+			setGeneticElements(geneticElements[0])
```
```suggestion
			setGeneticElements([...geneticElements, gene])
```
The reason your state change wasn't immediately triggering a re-render is because React checks for [referential equality](https://blog.bitsrc.io/understanding-referential-equality-in-react-a8fb3769be0) to listen for state change. 

In general it is good to view state objects as immutable and on any update to state, create a new object with the reflected changes. In this case you just need to create a new array copy with the new geneticElement appended to it. @zkdan Correct me if I made any errors here.

Yukthiw commented on Eplant/views/ChromosomeViewer/Viewer/GeneList.tsx:
```diff
@@ -0,0 +1,247 @@
+				arabidopsis,
+				selectedGene.aliases
+			)
+			geneticElements[0].push(gene)
```
```suggestion
```

Yukthiw commented on Eplant/views/ChromosomeViewer/Viewer/GeneList.tsx:
```diff
@@ -0,0 +1,247 @@
+	//------------------
+	// Helper Functions
+	//------------------
+	React.useEffect(() => {
```
```suggestion
	useEffect(() => {
```
Import individual exported members from react whenever possible

Yukthiw reviewed (changes requested):
Left a few comments largely related to updating the state of the gene list, will do a more formal review after you're finished with this.

y330 committed:
- added a feature that draws indicators on the chromosome for every gene in collection, but was creating duplicate indicators and had bug so its commented out
- fixed the radius of the small centormeric layers on chromsoome 3

y330 committed:
Finally the gene indicators are working for all geneticElements
- still has some bugs though
- changed the placement of the text and lines so that its dyanamic, based on the scale, to mirror eplant 2 functionality
- TODO:
add code comments where needed and clean up code
implement ESLint

y330 committed:
palcement fix for active gene indicator

y330 committed:
-cleaned up the code and made it adhere to project formatting standards
-added selectedGeneHistory to jotai state to make multiple geneInfoPopups be possible - currently not working
-changed name of geneIndicator to geneAnnotation

y330 committed:
added rudamentary support to view the poplar species
- dynamic gap spacing between chromosomes
- dynamic border radius of chromosomes

y330 committed:
added titlebar with zoom controls
switched mapinteractioncss library to react-zoomable-ui library for zoom, to better match zoom functionilty of original eplant site

y330 committed:
- replaced titlebar with ZoomControls, a floating button group
	- combined reset button and scale indicator text
- streamlined activeGeneAnnotation into the main loop to draw all geneAnnotations
- removed heatmap action
- cleaned up state imports in some files

y330 committed:
- ZoomControls: now shows when maximum and minimum zoom is reached in the tooltips
- Chromosome: added logic when drawing GeneAnnotations that checks if the geneAnnotation object actually corrosponds to an gene in a collection
	+ this is to prevent GeneAnnotations from staying on the screen after the corrosponding gene
	has been removed from the sidebar

y330 committed:
- moved the geneannotation logic to chromosome.tsx component
- got rid of activeGeneAnnotation prop
- fixed bug where gene annotations dont load when component is refreshed, now they do but it takes a few seconds

y330 committed:
Cleaned up code
fixed gene annotations not being too far away from the chromosome when very zoomed in
fixed snackbar placement

y330 committed:
cleanup

y330 committed:
ready for pr

Yukthiw commented on Eplant/views/ChromosomeViewer/Viewer/GeneInfoPopup.tsx:
```diff
@@ -0,0 +1,162 @@
+      const newGeneticElements = geneticElements[0]
+      newGeneticElements.push(geneticElement)
+      setGeneticElements(newGeneticElements)
+      setActiveGeneId(geneticElement.id)
```
```suggestion
      setGeneticElements([...geneticElements, geneticElement])
      setActiveGeneId(geneticElement.id)
```
This will update the state with a new memory reference which will trigger a refresh. With this change you can also remove the `activeGeneId` from the dependency array of the Collections component (I can't think of an instance where the list of genes needs updating on active gene change but I could be wrong on this).

Yukthiw commented on Eplant/main.tsx:
```diff
@@ -10,15 +10,15 @@ import './css/index.css'
 function RootApp() {
   return (
-    <React.StrictMode>
+    // <React.StrictMode>
```
Is there a reason that this is commented out?

Yukthiw commented on Eplant/views/ChromosomeViewer/Viewer/GeneInfoPopup.tsx:
```diff
@@ -0,0 +1,162 @@
+  const [gene, setGene] = useState<GeneItem>(props.gene)
+  // Global State
+  const [activeGeneId, setActiveGeneId] = useActiveGeneId()
+  const geneticElements = useGeneticElements()
```
```suggestion
  const [geneticElements, setGeneticElements] = useGeneticElements()
```

All of the Jotai atom hooks work like useStates

y330 replied:
when i change the geneticelements atom hook to useState format the geneInfoPopup  cannot add genes into the sidebar anymore

Yukthiw commented on Eplant/UI/LeftNav/Collections.tsx:
```diff
@@ -388,7 +393,11 @@ export function Collections(props: {
+  }, [
+    genes,
+    collections,
+    activeGeneId, // new
```
```suggestion
```

Yukthiw commented on Eplant/UI/LeftNav/Collections.tsx:
```diff
@@ -349,6 +353,7 @@ export function Collections(props: {
   onSelectGene?: (gene: GeneticElement) => void
   selectedGene?: string
 }) {
+  const [activeGeneId] = useActiveGeneId() // new
```
```suggestion
```

Yukthiw commented on Eplant/views/ChromosomeViewer/Viewer/Chromosome.tsx:
```diff
@@ -0,0 +1,380 @@
+  >([])
+
+  // Global State
+  const [geneticElements] = useGeneticElements()
+  const [collections] = useCollections()
```
```suggestion
  const [geneticElements, setGeneticElements] = useGeneticElements()
  const [collections, setCollections] = useCollections()
```

Yukthiw commented on Eplant/views/ChromosomeViewer/Viewer/Chromosome.tsx:
```diff
@@ -0,0 +1,380 @@
+  }, [])
+
+  //------------------
+  // Helper Functions
```
Would be a bit easier to read if you moved these helpers into a separate file

Yukthiw commented on Eplant/views/ChromosomeViewer/Viewer/GeneInfoPopup.tsx:
```diff
@@ -0,0 +1,162 @@
+  // Global State
+  const [activeGeneId, setActiveGeneId] = useActiveGeneId()
+  const geneticElements = useGeneticElements()
+  const setGeneticElements = useSetGeneticElements()
```
```suggestion
```

Can just use the setter function returned by `useGeneticElements()`

Yukthiw commented on Eplant/views/ChromosomeViewer/Viewer/GeneList.tsx:
```diff
@@ -0,0 +1,140 @@
+//----------
+const GeneList: FC<GeneListProps> = ({ id, start, end, anchorOrigin }) => {
+  // gene list
+  const [geneList, setGeneList] = useState<GeneItem[]>([
```
I think you can just initialize this to an empty array, else it can be a bit confusing to debug in the future where this list contains some dummy data.

Yukthiw commented on Eplant/views/ChromosomeViewer/Viewer/Viewer.tsx:
```diff
@@ -0,0 +1,52 @@
+//----------
+const Viewer: FC<ViewerProps> = ({ chromosomes, scale }) => {
+  return (
+    <div
```
nit: If possible try and use MUI boxes to keep things consistent

y330 replied:
using Box on line 23 messes up the spacing of the chromosomes

Yukthiw commented on Eplant/views/ChromosomeViewer/index.tsx:
```diff
@@ -0,0 +1,182 @@
+    gene: GeneticElement | null,
+    loadEvent: (progress: number) => void
+  ) {
+    let chromosomeViewData: ChromosomeItem[] = [
+      {
+        id: 'Chr1',
+        name: 'Chr 1',
+        size: 30427671,
+        centromeres: [],
+      },
+    ]
```
```suggestion
```

No need to initialize this here, can just type the `chromosomeViewData` below

Yukthiw commented on Eplant/views/ChromosomeViewer/index.tsx:
```diff
@@ -0,0 +1,182 @@
+      poplar ? '_poplar' : ''
+    }/cgi-bin/chromosomeinfo.cgi?species=${species}`
+
+    chromosomeViewData = await fetch(url)
```
```suggestion
    const chromosomeViewData: ChromosomeItem[] = await fetch(url)
```

Yukthiw commented on Eplant/views/ChromosomeViewer/Viewer/GeneAnnotation.tsx:
```diff
@@ -0,0 +1,81 @@
+      {/* GENE ID INDICATOR LINE */}
+      <line
+        x1={gene.strand == '+' ? 18 : 8}
+        y1={
```
On my local there seems to be a slight offset in the y-direction of the gene annotation lines:

<img width="355" alt="image" src="https://github.com/user-attachments/assets/09b19ee8-a960-42c4-9702-f848e563f08c">

It seems to be zoom agnostic so might just require a small adjustment. Let me know if this is just on my end

Yukthiw reviewed (changes requested):
Looks great Yonah! Thanks for doing all of this work. A small comments, largely to do with how state is managed. I noticed thing related to the y-coordinate of the geneAnnotation lines but hopefully this should be a relatively straightforward fix. 

I primarily looked at logic so if @zkdan you could take a look at some of the styling related changes that would be great (although it looks good to me).

y330 committed:
Merge branch 'staging' into ChromosomeView-Yonah

y330 committed:
removed uslayouteffect from index.tsx, apparently having snackbar component is enough to draw gene annotations after gene reload

y330 committed:
Merge branch 'ChromosomeView-Yonah' of https://github.com/y330/ePlant into ChromosomeView-Yonah

y330 committed:
applied suggested 1

y330 committed:
applied suggestion 2

y330 committed:
suggestion 3

y330 committed:
suggestion 4

y330 committed:
fixed slight offset in y-coordinate of  gene annotations

y330 committed:
READY FOR MERGE

y330 committed:
removed unneccesary logging lines - ACTUALLY READY FOR MERGING

Yukthiw commented on Eplant/views/ChromosomeViewer/Viewer/Chromosome.tsx:
```diff
@@ -0,0 +1,380 @@
+    const poplar = false
+    const species = poplar ? 'Populus_trichocarpa' : 'Arabidopsis_thaliana'
+    let newGeneAnnotationArray: GeneAnnotationItem[] = []
+    geneticElements.map((gene) => {
```
It seems that you are firing off an api call regardless of whether the geneticElement information has been stored already, and this is happening on every chromosome. I think this is logic that should occur only once at the top level component (the Viewer component) and the viewer can pass the relevant GeneItems to each chromosome. Ideally you would also only fire off an api call only for the new gene and not on every gene loaded (maintaining a set of gene ids should make this easy).

Let me know if this makes sense or if you'd like to hop on a quick call sometime to go over this.

**edit**

I'll leave this comment for visibility purposes but on second thought, we can leave this for another PR since we'll need to address the un-needed re-renders as well.

y330 replied:
Yeah, that makes sense, it can definitely be moved to the top level.

Yukthiw commented:
@y330 Thanks for addressing most of the comments, just a couple things before we get to merging this.

If you could move the helper functions in Chromosome.tsx to another file ("utils.tsx" or whatever) that would help with readability a lot. 

I made a comment about how we are handling the fetching of gene data, right now we are making a ton of api requests every time a new gene is added. Cutting down on these should improve performance by a lot but I think this can be left for another PR, just leave the comment there for reference.

Finally, before merging run `npx prettier Eplant --write` and push the changes to fix formatting issues (in the future adding the format on save feature to your vscode makes things easy)

y330 committed:
Moved Chromosome.tsx helper functions to new file, other minor changes. ran npx prettier Eplant --write

y330 commented:
@Yukthiw I moved the helper functions to a new file, and ran npx prettier Eplant --write. I am still not a contributor yet, so i cannot merge the PR

y330 committed:
READY FOR  MERGE

y330 committed:
ran npx prettier Eplant --write

Yukthiw reviewed (approved):
LGTM, thanks Yonah!
